### PR TITLE
Introduce accurate formulas for determinant and inverse of small matrices

### PR DESCRIPTION
**Pull request by IgorBaratta**
xtensor and numpy compute the determinant using LU decomposition.
This PR introduces formulas that are more accurate and faster for matrices up to 3x3.

Impact on performance:
```python3
mesh = dolfinx.UnitCubeMesh(comm, 100, 100, 100, CellType.tetrahedron)
V = dolfinx.FunctionSpace(mesh, ("N1curl", 1))
```

| | Main | This branch |
| :---         |     :---:      |          :---: |
| `compute_point_values`   | 21.30 s    | 17.13 s   |
| `interpolate`    | 47.74 s     | 32.91 s    |

```python3
mesh = dolfinx.UnitCubeMesh(comm, 100, 100, 100, CellType.hexahedron)
V = dolfinx.FunctionSpace(mesh, ("NCE", 1))
```
| | Main | This branch |
| :---         |     :---:      |          :---: |
| `compute_point_values`   | 19.59 s| 15.64 s  |
| `interpolate`    |  58.31 s      | 16.42 s  |

```python3
mesh = dolfinx.UnitCubeMesh(comm, 100, 100, 100, CellType.Tetrahedron)
V = dolfinx.FunctionSpace(mesh, ("Lagrange", 1))
```
| | Main | This branch |
| :---         |     :---:      |          :---: |
| `compute_point_values`   |  25.09  s   | 13.03 s  |
| `interpolate`    |  10.69  s    | 10.70 s |

### CODE REVIEW

File: cpp/dolfinx/common/linalg.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    B(0, 1) = difference_of_products(A(0, 2), A(0, 1), A(2, 2), A(2, 1)) * idet;
+    B(0, 2) = difference_of_products(A(0, 1), A(0, 2), A(1, 1), A(1, 2)) * idet;
+    B(1, 1) = difference_of_products(A(0, 0), A(0, 2), A(2, 0), A(2, 2)) * idet;
+    B(1, 2) = difference_of_products(A(1, 0), A(0, 0), A(1, 2), A(0, 2)) * idet;
+    B(2, 1) = difference_of_products(A(2, 0), A(0, 0), A(2, 1), A(0, 1)) * idet;
+    B(2, 2) = difference_of_products(A(0, 0), A(1, 0), A(0, 1), A(1, 1)) * idet;
+    break;
+  }
+  default:
+    throw std::runtime_error("linalg::inv is not implement for "
```
Comment (chrisrichardson) [inline cpp/dolfinx/common/linalg.h:104]: implemented

Comment (IgorBaratta) [reply cpp/dolfinx/common/linalg.h]: fixed

**Comment by garth-wells** [comment]
I think the namespace `linalg` is potentially confusing (confusing with `xt::linalg`, and confusing with `dolfinx::la`). Maybe `math` would be better?

**Comment by IgorBaratta** [comment]
> I think the namespace `linalg` is potentially confusing (confusing with `xt::linalg`, and confusing with `dolfinx::la`). Maybe `math` would be better?

I agree. 
Renamed it to `dolfinx::math`.

File: cpp/dolfinx/common/math.h [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace dolfinx::math
+{
+
+/// Kahan’s method to compute x = ad − bc with fused multiply-adds.
+/// The absolute error is bounded by 1.5 ulps, units of least precision.
+template <typename T>
+inline T difference_of_products(T a, T b, T c, T d) noexcept
+{
+  double w = b * c;
```
Comment (garth-wells) [inline cpp/dolfinx/common/math.h:19]: Should this be `T` rather than `double`?

File: cpp/dolfinx/common/math.h [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+/// Compute the inverse of a square matrix A and assign the result to a
+/// preallocated matrix B.
+/// @warning This function does not check if A is invertible!
+template <typename Matrix>
+void inv(const Matrix& A, Matrix& B)
+{
+  using value_type = typename Matrix::value_type;
+  value_type idet = 0;
```
Comment (garth-wells) [inline cpp/dolfinx/common/math.h:66]: Remove this.

File: cpp/dolfinx/common/math.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  const int nrows = A.shape(0);
+  switch (nrows)
+  {
+  case 1:
+    B(0, 0) = 1 / A(0, 0);
+    break;
+  case 2:
+  {
+    idet = 1. / det(A);
```
Comment (garth-wells) [inline cpp/dolfinx/common/math.h:75]: `value_type idet = 1. / det(A);`

File: cpp/dolfinx/common/math.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  }
+  case 3:
+  {
+    value_type w0 = difference_of_products(A(1, 1), A(1, 2), A(2, 1), A(2, 2));
+    value_type w1 = difference_of_products(A(1, 0), A(1, 2), A(2, 0), A(2, 2));
+    value_type w2 = difference_of_products(A(1, 0), A(1, 1), A(2, 0), A(2, 1));
+    value_type w3 = difference_of_products(A(0, 0), A(0, 1), w1, w0);
+    value_type det = std::fma(A(0, 2), w2, w3);
+    assert(det != 0.);
+    idet = 1 / det;
```
Comment (garth-wells) [inline cpp/dolfinx/common/math.h:90]: value_type idet = 1. / det(A);

File: cpp/dolfinx/common/math.h [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                             + std::to_string(A.shape(0)) + "x"
+                             + std::to_string(A.shape(1)) + " matrices.");
+  }
+}
+
+/// Compute the inverse of a square matrix A and assign the result to a
+/// preallocated matrix B.
+/// @warning This function does not check if A is invertible!
+template <typename Matrix>
+void inv(const Matrix& A, Matrix& B)
```
Comment (garth-wells) [inline cpp/dolfinx/common/math.h:63]: I think we should we allow `A` and `B` to have different types (e.g., one might be `xt::xtensor_fixed` and the other `xt::xtensor`).

**Review by garth-wells — CHANGES_REQUESTED** [review]

**Review by garth-wells — APPROVED** [review]

**Pull request merged**